From ae92032db1a532225e4b43277bec7f8c3ae3a6cf Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Wed, 26 Jun 2024 21:10:46 +0200
Subject: [PATCH] Update to Dogecoin block subsidy (#14)

Dogecoin's block subsidy works very differently, and had a few changes over time.

Keep the subsidy mechanism on Regtest so we don't break the tests.

(cherry picked from commit 0627acfef5a82ad2032fa4a401d6775eafe2a4f0)

diff --git a/src/bench/duplicate_inputs.cpp b/src/bench/duplicate_inputs.cpp
index 2d8ec33b2e4a7a56659b44e9da00a7611d50d459..e7521d26a240f29637d4f255d4ed76249fb8ab49 100644
--- a/src/bench/duplicate_inputs.cpp
+++ b/src/bench/duplicate_inputs.cpp
@@ -42,7 +42,8 @@ static void DuplicateInputs(benchmark::Bench &bench) {
     coinbaseTx.vin[0].prevout = COutPoint();
     coinbaseTx.vout.resize(1);
     coinbaseTx.vout[0].scriptPubKey = SCRIPT_PUB;
-    coinbaseTx.vout[0].nValue = GetBlockSubsidy(nHeight, consensusParams);
+    coinbaseTx.vout[0].nValue =
+        GetBlockSubsidy(nHeight, consensusParams, pindexPrev->GetBlockHash());
     coinbaseTx.vin[0].scriptSig = CScript() << nHeight << OP_0;
 
     naughtyTx.vout.resize(1);
diff --git a/src/index/coinstatsindex.cpp b/src/index/coinstatsindex.cpp
index 470cc17d3e3c333e66a310519d86c4c5a9e647cc..f164f1c1fc5f57e60d81a69904e8efc24b8465d1 100644
--- a/src/index/coinstatsindex.cpp
+++ b/src/index/coinstatsindex.cpp
@@ -114,8 +114,8 @@ CoinStatsIndex::CoinStatsIndex(std::unique_ptr<interfaces::Chain> chain,
 bool CoinStatsIndex::WriteBlock(const CBlock &block,
                                 const CBlockIndex *pindex) {
     CBlockUndo block_undo;
-    const Amount block_subsidy{
-        GetBlockSubsidy(pindex->nHeight, Params().GetConsensus())};
+    const Amount block_subsidy{GetBlockSubsidy(
+        pindex->nHeight, Params().GetConsensus(), block.hashPrevBlock)};
     m_total_subsidy += block_subsidy;
 
     // Ignore genesis block
@@ -440,8 +440,8 @@ bool CoinStatsIndex::ReverseBlock(const CBlock &block,
     CBlockUndo block_undo;
     std::pair<BlockHash, DBVal> read_out;
 
-    const Amount block_subsidy{
-        GetBlockSubsidy(pindex->nHeight, Params().GetConsensus())};
+    const Amount block_subsidy{GetBlockSubsidy(
+        pindex->nHeight, Params().GetConsensus(), block.hashPrevBlock)};
     m_total_subsidy -= block_subsidy;
 
     // Ignore genesis block
diff --git a/src/node/miner.cpp b/src/node/miner.cpp
index 4b3257d48752189a1800587446bf5b4f20876563..f7321c8d68e030d9ace75784dd78840df041c18b 100644
--- a/src/node/miner.cpp
+++ b/src/node/miner.cpp
@@ -171,7 +171,7 @@ BlockAssembler::CreateNewBlock(const CScript &scriptPubKeyIn) {
     coinbaseTx.vout.resize(1);
     coinbaseTx.vout[0].scriptPubKey = scriptPubKeyIn;
     coinbaseTx.vout[0].nValue =
-        blockFitter.nFees + GetBlockSubsidy(nHeight, consensusParams);
+        blockFitter.nFees + GetBlockSubsidy(nHeight, consensusParams, pindexPrev->GetBlockHash());
     coinbaseTx.vin[0].scriptSig = CScript() << nHeight << OP_0;
 
     const Amount blockReward = coinbaseTx.vout[0].nValue;
diff --git a/src/rpc/blockchain.cpp b/src/rpc/blockchain.cpp
index ed3047776562fed3dad8364b20eafde7174c7f87..2ced43c9f9f85b780662dcb67537551d65cd043c 100644
--- a/src/rpc/blockchain.cpp
+++ b/src/rpc/blockchain.cpp
@@ -2272,7 +2272,8 @@ static RPCHelpMan getblockstats() {
                            mintxsize == blockMaxSize ? 0 : mintxsize);
             ret_all.pushKV("outs", outputs);
             ret_all.pushKV("subsidy", GetBlockSubsidy(pindex.nHeight,
-                                                      chainman.GetConsensus()));
+                                                      chainman.GetConsensus(),
+                                                      block.hashPrevBlock));
             ret_all.pushKV("time", pindex.GetBlockTime());
             ret_all.pushKV("total_out", total_out);
             ret_all.pushKV("total_size", total_size);
diff --git a/src/test/policy_block_tests.cpp b/src/test/policy_block_tests.cpp
index 0173ec6c734a3ee3bc0037f228c117af776f290c..eb31f9c20dc5c5a3f871ba577fab2179a86dad38 100644
--- a/src/test/policy_block_tests.cpp
+++ b/src/test/policy_block_tests.cpp
@@ -59,7 +59,8 @@ BOOST_AUTO_TEST_CASE(policy_minerfund) {
     CBlockIndex &lastBlockIndexRef = blocks.back();
 
     const Amount blockReward =
-        GetBlockSubsidy(lastBlockIndexRef.nHeight, consensusParams);
+        GetBlockSubsidy(lastBlockIndexRef.nHeight, consensusParams,
+                        lastBlockIndexRef.pprev->GetBlockHash());
 
     auto checkMinerFundPolicy = [&](CBlock block, const CBlockIndex &blockIndex,
                                     bool expected) {
diff --git a/src/test/validation_tests.cpp b/src/test/validation_tests.cpp
index fb3ed01ae9032977c6ba7a69d1f73e3ea0d0c6ac..f5bb2b2b87ba0a0b246e9f593eb2be212590a719 100644
--- a/src/test/validation_tests.cpp
+++ b/src/test/validation_tests.cpp
@@ -27,11 +27,178 @@
 
 #include <cstdint>
 #include <cstdio>
-#include <string>
+#include <tuple>
 #include <vector>
 
 BOOST_FIXTURE_TEST_SUITE(validation_tests, TestingSetup)
 
+// First 20 blocks on the Dogecoin network, with their coinbase subsidy
+const std::vector<std::tuple<std::string, int64_t>> DOGECOIN_BLOCKS = {
+    // Genesis block has 88 DOGE blockreward, but not used in the tests here.
+    {"1a91e3dace36e2be3bf030a65679fe821aa1d6ef92e7c9902eb318182c355691", 88},
+    {"82bc68038f6034c0596b6e313729793a887fded6e92a31fbdf70863f89d9bea2", 68416},
+    {"ea5380659e02a68c073369e502125c634b2fb0aaf351b9360c673368c4f20c96",
+     729752},
+    {"76f80a8a81e6f6669d340651723b874f97395c4dbda200f8b024df4c6566a92c", 11183},
+    {"df363f95151d8c38b1cf0ee8375d571c9a869d9e37489ba058982ace19bcdee4",
+     324378},
+    {"f21dc70cb44c180261e31a222202678602d605e7697332cb2395386fa309ad3b",
+     262711},
+    {"f34986a114a2f58f48ce5593e5e6006666243fb003a2ffa489d980d8de825428",
+     863413},
+    {"3ca7e813da5c72b0817c4d4789cd4896f49baf5e40f67a59a63a63ea2498d604",
+     141339},
+    {"ed6b216e69b57915eda3a43036016c5667b35d61606f1c61ba10fb17c0e7a063",
+     857035},
+    {"0ddd48852cb794c7534841a2cd3507e40255b1707fac14d65d16cc791672e5e2",
+     515758},
+    {"31bf9377a7f52b97bc7ec001a107b3c3c0f7ff629767c2bcc4e27873dad5fc21",
+     890107},
+    {"45b727adf6b098223d8360332f51256cf00dd9a3501cf09d50767264e733fe72",
+     241677},
+    {"30fd7677bc28675e8663feef4f4b6ceab936407a3a365e4af4cc7e14b62bdd3d",
+     279377},
+    {"4a991ce1ca12b6160776a978f5ac0635b9d497c2516fdfa37894a4d7bfe2f66e",
+     909605},
+    {"f95c467d4ed4c0082b7da6b5fbffd60e33ab55ad0541358f945d7243674058b0",
+     348510},
+    {"9ff299652aba8a4de7f8edfa42500de0e47525486674c23ad68d3147d19514a1",
+     837643},
+    {"fbc2620f0cb7f490d3cba09ddadb9c79fca7187180d7cf9053537e3bc872c1fb",
+     214794},
+    {"0c120ab190655673a709bc92ad86f80dc1cd9f11f9e0f09ebc5e6a3058b73002",
+     434206},
+    {"da0e2362cc1d1cd48c8eb70e578c97f00d9a530985ba36027eb7e3fba98c74ae",
+     190398},
+    {"7f34d92c06a2b38cac860d62026e716c9a73a759363891b1bf7d0cd465c6acba",
+     653360},
+};
+
+BOOST_AUTO_TEST_CASE(subsidy_mainnet_first_20_blocks) {
+    const auto chainParams =
+        CreateChainParams(*m_node.args, CBaseChainParams::MAIN);
+    const Consensus::Params &params = chainParams->GetConsensus();
+    for (int32_t height = 1; height < (int32_t)DOGECOIN_BLOCKS.size();
+         ++height) {
+        Amount subsidy = GetBlockSubsidy(
+            height, params, uint256S(std::get<0>(DOGECOIN_BLOCKS[height - 1])));
+        BOOST_CHECK_EQUAL(subsidy, std::get<1>(DOGECOIN_BLOCKS[height]) * COIN);
+    }
+}
+
+BOOST_AUTO_TEST_CASE(subsidy_first_100k_test) {
+    const auto chainParams =
+        CreateChainParams(*m_node.args, CBaseChainParams::MAIN);
+    const Consensus::Params &params = chainParams->GetConsensus();
+    Amount nSum;
+    arith_uint256 prevHash;
+
+    for (int32_t nHeight = 0; nHeight <= 100000; nHeight++) {
+        Amount nSubsidy =
+            GetBlockSubsidy(nHeight, params, ArithToUint256(prevHash));
+        BOOST_CHECK(MoneyRange(nSubsidy));
+        BOOST_CHECK(nSubsidy <= 1000000 * COIN);
+        nSum += nSubsidy;
+        // Use nSubsidy to give us some variation in previous block hash,
+        // without requiring full block templates
+        // Must shift 200 bits because only bits 200-227 are used for the seed
+        prevHash += arith_uint256(nSubsidy / SATOSHI) << 200;
+    }
+
+    const Amount expected = int64_t(49555514013LL) * COIN;
+    BOOST_CHECK_EQUAL(expected, nSum);
+}
+
+BOOST_AUTO_TEST_CASE(subsidy_100k_145k_test) {
+    const auto chainParams =
+        CreateChainParams(*m_node.args, CBaseChainParams::MAIN);
+    const Consensus::Params &params = chainParams->GetConsensus();
+    Amount nSum;
+    arith_uint256 prevHash;
+
+    for (int32_t nHeight = 100000; nHeight <= 145000; nHeight++) {
+        Amount nSubsidy =
+            GetBlockSubsidy(nHeight, params, ArithToUint256(prevHash));
+        BOOST_CHECK(MoneyRange(nSubsidy));
+        BOOST_CHECK(nSubsidy <= 500000 * COIN);
+        nSum += nSubsidy;
+        // Use nSubsidy to give us some variation in previous block hash,
+        // without requiring full block templates
+        // Must shift 200 bits because only bits 200-227 are used for the seed
+        prevHash += arith_uint256(nSubsidy / SATOSHI) << 200;
+    }
+
+    const Amount expected = int64_t(11029457731LL) * COIN;
+    BOOST_CHECK_EQUAL(expected, nSum);
+}
+
+// Check the simplified rewards after block 145,000
+BOOST_AUTO_TEST_CASE(subsidy_post_145k_test) {
+    const auto chainParams =
+        CreateChainParams(*m_node.args, CBaseChainParams::MAIN);
+    const Consensus::Params &params = chainParams->GetConsensus();
+    const uint256 prevHash;
+
+    for (int32_t nHeight = 145000; nHeight < 600000; nHeight++) {
+        Amount nSubsidy = GetBlockSubsidy(nHeight, params, prevHash);
+        Amount nExpectedSubsidy = (500000 >> (nHeight / 100000)) * COIN;
+        BOOST_CHECK(MoneyRange(nSubsidy));
+        BOOST_CHECK_EQUAL(nSubsidy, nExpectedSubsidy);
+    }
+
+    // Test reward at 600k+ is constant
+    Amount nConstantSubsidy = GetBlockSubsidy(600000, params, prevHash);
+    BOOST_CHECK_EQUAL(nConstantSubsidy, 10000 * COIN);
+
+    nConstantSubsidy = GetBlockSubsidy(700000, params, prevHash);
+    BOOST_CHECK_EQUAL(nConstantSubsidy, 10000 * COIN);
+}
+
+// This test is broken; it is copied from the original dogecoin test suite to
+// prevent divergence. The test tweaks prevHash, but the bits determining the
+// seed are untouched, such that the seed is always 0, resulting in no test
+// coverage for the seed extraction code.
+BOOST_AUTO_TEST_CASE(broken_dogecoin_subsidy_first_100k_test) {
+    const auto chainParams =
+        CreateChainParams(*m_node.args, CBaseChainParams::MAIN);
+    const Consensus::Params &params = chainParams->GetConsensus();
+    Amount nSum;
+    arith_uint256 prevHash;
+
+    for (int nHeight = 0; nHeight <= 100000; nHeight++) {
+        Amount nSubsidy =
+            GetBlockSubsidy(nHeight, params, ArithToUint256(prevHash));
+        BOOST_CHECK(MoneyRange(nSubsidy));
+        BOOST_CHECK(nSubsidy <= 1000000 * COIN);
+        nSum += nSubsidy;
+        prevHash += nSubsidy / SATOSHI;
+    }
+
+    const Amount expected = int64_t(54894174438LL) * COIN;
+    BOOST_CHECK_EQUAL(expected, nSum);
+}
+
+// This test is broken, see broken_dogecoin_subsidy_first_100k_test
+BOOST_AUTO_TEST_CASE(broken_dogecoin_subsidy_100k_145k_test) {
+    const auto chainParams =
+        CreateChainParams(*m_node.args, CBaseChainParams::MAIN);
+    const Consensus::Params &params = chainParams->GetConsensus();
+    Amount nSum;
+    arith_uint256 prevHash;
+
+    for (int nHeight = 100000; nHeight <= 145000; nHeight++) {
+        Amount nSubsidy =
+            GetBlockSubsidy(nHeight, params, ArithToUint256(prevHash));
+        BOOST_CHECK(MoneyRange(nSubsidy));
+        BOOST_CHECK(nSubsidy <= 500000 * COIN);
+        nSum += nSubsidy;
+        prevHash += nSubsidy / SATOSHI;
+    }
+
+    const Amount expected = int64_t(12349960000LL) * COIN;
+    BOOST_CHECK_EQUAL(expected, nSum);
+}
+
 static void TestBlockSubsidyHalvings(const Consensus::Params &consensusParams) {
     int maxHalvings = 64;
     Amount nInitialSubsidy = 50 * COIN;
@@ -39,29 +206,29 @@ static void TestBlockSubsidyHalvings(const Consensus::Params &consensusParams) {
     // for height == 0
     Amount nPreviousSubsidy = 2 * nInitialSubsidy;
     BOOST_CHECK_EQUAL(nPreviousSubsidy, 2 * nInitialSubsidy);
-    for (int nHalvings = 0; nHalvings < maxHalvings; nHalvings++) {
-        int nHeight = nHalvings * consensusParams.nSubsidyHalvingInterval;
-        Amount nSubsidy = GetBlockSubsidy(nHeight, consensusParams);
+    for (int32_t nHalvings = 0; nHalvings < maxHalvings; nHalvings++) {
+        int32_t nHeight = nHalvings * consensusParams.nSubsidyHalvingInterval;
+        Amount nSubsidy = GetBlockSubsidy(nHeight, consensusParams, uint256());
         BOOST_CHECK(nSubsidy <= nInitialSubsidy);
         BOOST_CHECK_EQUAL(nSubsidy, nPreviousSubsidy / 2);
         nPreviousSubsidy = nSubsidy;
     }
     BOOST_CHECK_EQUAL(
         GetBlockSubsidy(maxHalvings * consensusParams.nSubsidyHalvingInterval,
-                        consensusParams),
+                        consensusParams, uint256()),
         Amount::zero());
 }
 
 static void TestBlockSubsidyHalvings(int nSubsidyHalvingInterval) {
     Consensus::Params consensusParams;
+    consensusParams.fPowNoRetargeting = true;
     consensusParams.nSubsidyHalvingInterval = nSubsidyHalvingInterval;
     TestBlockSubsidyHalvings(consensusParams);
 }
 
 BOOST_AUTO_TEST_CASE(block_subsidy_test) {
-    const auto chainParams = CreateChainParams(*m_node.args, ChainType::MAIN);
-    // As in main
-    TestBlockSubsidyHalvings(chainParams->GetConsensus());
+    // As in Bitcoin
+    TestBlockSubsidyHalvings(210000);
     // As in regtest
     TestBlockSubsidyHalvings(150);
     // Just another interval
@@ -69,10 +236,12 @@ BOOST_AUTO_TEST_CASE(block_subsidy_test) {
 }
 
 BOOST_AUTO_TEST_CASE(subsidy_limit_test) {
-    const auto chainParams = CreateChainParams(*m_node.args, ChainType::MAIN);
+    Consensus::Params params;
+    params.fPowNoRetargeting = true;
+    params.nSubsidyHalvingInterval = 210000; // Bitcoin
     Amount nSum = Amount::zero();
     for (int nHeight = 0; nHeight < 14000000; nHeight += 1000) {
-        Amount nSubsidy = GetBlockSubsidy(nHeight, chainParams->GetConsensus());
+        Amount nSubsidy = GetBlockSubsidy(nHeight, params, uint256());
         BOOST_CHECK(nSubsidy <= 50 * COIN);
         nSum += 1000 * nSubsidy;
         BOOST_CHECK(MoneyRange(nSum));
diff --git a/src/validation.cpp b/src/validation.cpp
index b8bf3a9bc9d4bc93730524e7f057e0cc227dfdeb..25fdc5d9d2771b45933053d68decacc51c72d139 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -75,6 +75,9 @@
 #include <thread>
 #include <tuple>
 
+#include <boost/random/mersenne_twister.hpp>
+#include <boost/random/uniform_int.hpp>
+
 using kernel::CCoinsStats;
 using kernel::CoinStatsHashType;
 using kernel::ComputeUTXOStats;
@@ -1485,17 +1488,46 @@ PackageMempoolAcceptResult ProcessNewPackage(Chainstate &active_chainstate,
     return result;
 }
 
-Amount GetBlockSubsidy(int nHeight, const Consensus::Params &consensusParams) {
+Amount GetBlockSubsidy(int nHeight, const Consensus::Params &consensusParams,
+                       uint256 prevHash) {
     int halvings = nHeight / consensusParams.nSubsidyHalvingInterval;
-    // Force block reward to zero when right shift is undefined.
-    if (halvings >= 64) {
-        return Amount::zero();
+
+    // On regtest, use the Bitcoin halving mechanism
+    if (consensusParams.fPowNoRetargeting) {
+        // Force block reward to zero when right shift is undefined.
+        if (halvings >= 64) {
+            return Amount::zero();
+        }
+
+        Amount nSubsidy = 50 * COIN;
+        // Subsidy is cut in half every 210,000 blocks which will occur
+        // approximately every 4 years.
+        return ((nSubsidy / SATOSHI) >> halvings) * SATOSHI;
     }
 
-    Amount nSubsidy = 50 * COIN;
-    // Subsidy is cut in half every 210,000 blocks which will occur
-    // approximately every 4 years.
-    return ((nSubsidy / SATOSHI) >> halvings) * SATOSHI;
+    if (!IsDigishieldEnabled(consensusParams, nHeight)) {
+        int64_t maxReward = (1000000 >> halvings) - 1;
+
+        // Old-style rewards derived from the previous block hash
+        // Extract bits 200 to 227
+        // 0000000XXXXXXX00000000000000000000000000000000000000000000000000
+        arith_uint256 prev = UintToArith256(prevHash);
+        prev <<= 28; // Remove leading 7 hex digits
+        prev >>= 256 - 28;
+        int64_t seed = (int64_t)prev.GetLow64();
+
+        boost::mt19937 gen(seed);
+        boost::uniform_int<> dist(1, maxReward);
+        int32_t rand = dist(gen);
+
+        return (1 + rand) * COIN;
+    } else if (nHeight < (6 * consensusParams.nSubsidyHalvingInterval)) {
+        // New-style constant rewards for each halving interval
+        return ((500000 * COIN / SATOSHI) >> halvings) * SATOSHI;
+    } else {
+        // Constant inflation
+        return 10000 * COIN;
+    }
 }
 
 CoinsViews::CoinsViews(DBParams db_params, CoinsViewOptions options)
@@ -2460,7 +2492,7 @@ bool Chainstate::ConnectBlock(const CBlock &block, BlockValidationState &state,
                  m_chainman.num_blocks_total);
 
     const Amount blockReward =
-        nFees + GetBlockSubsidy(pindex->nHeight, consensusParams);
+        nFees + GetBlockSubsidy(pindex->nHeight, consensusParams, pindex->pprev->GetBlockHash());
     if (block.vtx[0]->GetValueOut() > blockReward && state.IsValid()) {
         state.Invalid(
             BlockValidationResult::BLOCK_CONSENSUS, "bad-cb-amount",
@@ -2991,8 +3023,8 @@ bool Chainstate::ConnectTip(BlockValidationState &state,
             m_filterParkingPoliciesApplied.insert(blockhash);
 
             const Amount blockReward =
-                blockFees +
-                GetBlockSubsidy(pindexNew->nHeight, consensusParams);
+                blockFees + GetBlockSubsidy(pindexNew->nHeight, consensusParams,
+                                            pindexNew->pprev->GetBlockHash());
 
             std::vector<std::unique_ptr<ParkingPolicy>> parkingPolicies;
             parkingPolicies.emplace_back(std::make_unique<MinerFundPolicy>(
diff --git a/src/validation.h b/src/validation.h
index 3e4f03567b6e83fbfee36121a21229c189ef9074..b1b47ed9cf8bde191bb5c746fa2e79f0355776e5 100644
--- a/src/validation.h
+++ b/src/validation.h
@@ -152,7 +152,7 @@ public:
     uint64_t getExcessiveBlockSize() const { return excessiveBlockSize; }
 };
 
-Amount GetBlockSubsidy(int nHeight, const Consensus::Params &consensusParams);
+Amount GetBlockSubsidy(int nHeight, const Consensus::Params &consensusParams, uint256 prevHash);
 
 bool FatalError(kernel::Notifications &notifications,
                 BlockValidationState &state, const std::string &strMessage,
diff --git a/test/lint/lint-boost-dependencies.sh b/test/lint/lint-boost-dependencies.sh
index 3b020037399d8ef4b67fb9dd0ca0f88725d29b8a..5e54b1703816c24cb603e13fe0a60691374cf32b 100755
--- a/test/lint/lint-boost-dependencies.sh
+++ b/test/lint/lint-boost-dependencies.sh
@@ -19,6 +19,8 @@ EXPECTED_BOOST_INCLUDES=(
     boost/multi_index/ordered_index.hpp
     boost/multi_index/sequenced_index.hpp
     boost/multi_index_container.hpp
+    boost/random/mersenne_twister.hpp
+    boost/random/uniform_int.hpp
     boost/range/iterator.hpp
     boost/range/adaptor/sliced.hpp
     boost/signals2/connection.hpp
